src/main/java/htsjdk/io/AsyncWriterPool.java

```diff
@@ -0,0 +1,270 @@
+         * set {@code writeThreshold} and the previous task has been checked and is now null, the queue will be sent to
+         * the {@link AsyncWriterPool}'s executor for draining.
+         *
+         * @param item an item to enqueue for writing
+         * @throws RuntimeIOException if queue is already closed
+         * @throws RuntimeException   if the previous attempt to write failed, or if the attempt to add to the queue has
+         *                            been interrupted
+         */
+        @Override
+        public void write(final A item) {
+            if (this.isClosed) throw new RuntimeIOException("Attempt to add record to closed writer.");
+
+            this.nonBlockingCheckAndRethrow();
+
+            // Put new item in queue
+            try {
+                while (!this.isClosed && !this.queue.offer(item, 5, TimeUnit.SECONDS)) {
+                }
+            } catch (InterruptedException e) {
+                throw new RuntimeException("Exception while placing item in queue", e);
+            }
+
+            // If queue size is large enough, and the there is no last task
+            if (queue.size() >= this.writeThreshold && this.currentTask == null) {
+                this.drain();
+            }
+        }
+
+        /**
+         * Launch a task in the {@link AsyncWriterPool} to pull items from this writers queue, and write them, till the
+         * queue is empty.
+         * <p>
+         * If items are being added to the queue faster than they can be written, and the queue is unbounded, the queue
+         * may result in an {@code OutOfMemoryError}.
+         */
+        private void drain() {
+            // check the result of the previous task
+            this.currentTask = AsyncWriterPool.this.executor.submit(() -> {
+
+                // BlockingQueue is threadsafe and uses a different lock for both reading and writing
+                // ends of the queue, so this is safe and avoids an extra copy.
+                A item = this.queue.poll();
+                while (item != null) {
+                    this.writer.write(item);
+                    item = this.queue.poll();
+                }
+                return null;
+            });
+        }
+
+        /**
+         * Non-blocking close the the enclosed writer. This wraps a call to {@link PooledWriter#blockingClose()} and
+         * sends it to execute on the executor. {@code .isClosed} is immediately set to {@code true} on the current
+         * thread. The new thread will block until any ongoing writes are complete, drain any remaining enqueued items,
+         * check for exceptions, and then call close on the underlying writer.
+         *
+         * @return a future that can be waited on
+         * @throws IllegalStateException if isClosed is already true
+         */
+        private CompletableFuture<Void> nonBlockingClose() {
+            if (this.isClosed) throw new IllegalStateException("Writer is already closed");
```

Comment:
I tend to think that this should just return - I think it's totally reasonable that I as a user might first close the writers independently (they do implement closeable after all) and _then_ close the pool which is going to call these methods again.